### PR TITLE
[try_put_and_wait] Part 4: Add implementation of try_put_and_wait API for split_node

### PR DESCRIPTION
**Pull request by kboyarinov**
### Description 
Implement try_put_and_wait for split_node + test

Fixes # - _issue number(s) if exists_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [x] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [x] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### CODE REVIEW

File: include/oneapi/tbb/detail/_flow_graph_node_impl.h [hunk 1]
```diff
@@ -647,6 +647,18 @@ struct emit_element {
         check_task_and_spawn(g, last_task);
         return emit_element<N-1>::emit_this(g,t,p);
     }
+
+#if __TBB_PREVIEW_FLOW_GRAPH_TRY_PUT_AND_WAIT
+    template <typename TupleType, typename PortsType>
+    static graph_task* emit_this(graph& g, const TupleType& t, PortsType& p,
```
Comment (pavelkumbrasev) [inline include/oneapi/tbb/detail/_flow_graph_node_impl.h:653]: I'm not sure what this method supposed to do. Could you please explain?

File: test/tbb/test_split_node.cpp [hunk 2]
```diff
@@ -397,6 +397,83 @@ void test_deduction_guides() {
...
+        // and only when {f2 - 10} would be taken and executed, try_put_and_wait would be exitted
+        // All of the other tasks for start_work_items would be processed during wait_for_all()
+        CHECK_MESSAGE(processed_items1[check_index1++] == wait_message, "Unexpected items processing");
+
+        for (std::size_t i = new_work_items.size(); i != 0; --i) {
+            CHECK_MESSAGE(processed_items1[check_index1++] == new_work_items[i - 1], "Unexpected items processing");
+            CHECK_MESSAGE(processed_items2[check_index2++] == new_work_items[i - 1], "Unexpected items processing");
+        }
+
+        CHECK_MESSAGE(processed_items2[check_index2++] == wait_message, "Unexpected items processing");
```
Comment (pavelkumbrasev) [inline test/tbb/test_split_node.cpp:465]:
```suggestion
        CHECK_MESSAGE(processed_items1[check_index1++] == wait_message, "Unexpected items processing");
        CHECK_MESSAGE(processed_items2[check_index2++] == wait_message, "Unexpected items processing");

        for (std::size_t i = new_work_items.size(); i != 0; --i) {
            CHECK_MESSAGE(processed_items1[check_index1++] == new_work_items[i - 1], "Unexpected items processing");
            CHECK_MESSAGE(processed_items2[check_index2++] == new_work_items[i - 1], "Unexpected items processing");
        }
```

**Review by pavelkumbrasev — COMMENTED** [review]
Other than that LGTM!

File: include/oneapi/tbb/detail/_flow_graph_node_impl.h (hunk 1 above)
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_node_impl.h:653]: Sure. This function is a helper function to `split_node`.
Let's say the `split_node` takes a `tuple<Arg1, Arg2, ..., ArgN> and sends the corresponding item to each of N successors.
`split_node::output_ports()` is a tuple of N successors which is passed here (argument `t`).
`emit_this<N>` takes Nth element from the input tuple, calls `try_put_task` on the Nth output port and spawns the returned task. 
New function is just an overload for pre-existing `emit_this` sending `metainfo` together with the element.

File: test/tbb/test_split_node.cpp (hunk 2 above)
Comment (kboyarinov) [reply test/tbb/test_split_node.cpp:465]: Reverted - the proposed change is not valid. See the comment to the test for more detailed explanation.

**Pull request merged**